gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java

```diff
@@ -84,21 +84,21 @@ private boolean includeField(Field f, boolean serialize) {
   /** first element holds the default name */
   @SuppressWarnings("MixedMutabilityReturnType")
   private List<String> getFieldNames(Field f) {
+
+    String fieldName;
+    List<String> alternates;
     SerializedName annotation = f.getAnnotation(SerializedName.class);
     if (annotation == null) {
-      String name = fieldNamingPolicy.translateName(f);
-      return Collections.singletonList(name);
-    }
-
-    String serializedName = annotation.value();
-    String[] alternates = annotation.alternate();
-    if (alternates.length == 0) {
-      return Collections.singletonList(serializedName);
+      fieldName = fieldNamingPolicy.translateName(f);
+      alternates = fieldNamingPolicy.alternateNames(f);
+    } else {
+      fieldName = annotation.value();
+      alternates = Arrays.asList(annotation.alternate());
     }
 
-    List<String> fieldNames = new ArrayList<>(alternates.length + 1);
-    fieldNames.add(serializedName);
-    Collections.addAll(fieldNames, alternates);
+    List<String> fieldNames = new ArrayList<>(alternates.size() + 1);
```

Comment:
Although it isn't strictly necessary, I think the special case for `Collections.singletonList` should be preserved, because it is also the usual case.

Reply:
The special case with Collections.singletonList has been added back.